poetry: 1.4.2 -> 1.5.1

Dessix opened the pull request:
poetry: 1.4.2 -> 1.5.1

Dropped support for python-plugin-up which has not been updated past poetry 1.4.2 after 1.5.x's significant change from tomlkit to tomli.

Based largely upon work from the hence-abandoned [PR #233393](https://github.com/NixOS/nixpkgs/pull/233393).

Diff: https://github.com/python-poetry/poetry/compare/refs/tags/1.4.2...1.5.1

Changelogs:
1.4.2 -> 1.5.0: [python-poetry/poetry@refs/tags/1.5.0/CHANGELOG.md](https://github.com/python-poetry/poetry/blob/refs/tags/1.5.0/CHANGELOG.md?rgh-link-date=2023-07-04T10%3A34%3A42Z)
1.5.0 -> 1.5.1: [python-poetry/poetry@refs/tags/1.5.1/CHANGELOG.md](https://github.com/python-poetry/poetry/blob/refs/tags/1.5.1/CHANGELOG.md?rgh-link-date=2023-07-04T10%3A34%3A42Z)

Dropped support for [poetry-plugin-up](https://github.com/MousaZeidBaker/poetry-plugin-up/) as it has not been updated for compatibility with Poetry's new core and is apparently not taking pull requests at the moment.

###### Description of changes

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Dessix committed:
poetry: 1.4.2 -> 1.5.1

Based largely upon work from the hence-abandoned [PR #233393](https://github.com/NixOS/nixpkgs/pull/233393).

Diff: https://github.com/python-poetry/poetry/compare/refs/tags/1.4.2...1.5.1

Changelogs:
- 1.4.2 -> 1.5.0: [python-poetry/poetry@refs/tags/1.5.0/CHANGELOG.md](https://github.com/python-poetry/poetry/blob/refs/tags/1.5.0/CHANGELOG.md?rgh-link-date=2023-07-04T10%3A34%3A42Z)

- 1.5.0 -> 1.5.1: [python-poetry/poetry@refs/tags/1.5.1/CHANGELOG.md](https://github.com/python-poetry/poetry/blob/refs/tags/1.5.1/CHANGELOG.md?rgh-link-date=2023-07-04T10%3A34%3A42Z)

mweinelt commented:
> poetry: 1.4.2 -> 1.5.1

All commit messages lack changelogs, as required by CONTRIBUTING.md.

> poetry-core: 1.5.1 -> 1.6.1

That's too big a change for master. This is a blocker currently, hence marking it as a draft. You could use an override for the time being.

> poetry-plugin-export: 1.3.1 -> 1.4.0
> python310Packages.poetry-plugin-export: 1.3.1 -> 1.4.0
> python311Packages.poetry-plugin-export: 1.3.1 -> 1.4.0

Every changed attribute should roughly receive its own commit. So one commit per plugin for example.

SuperSandro2000 commented:
> I was avoiding this because that left multiple commits wherein each one was broken; if you'd prefer, I can indeed split them out.

We do one commit per package update even if between the commits something is broken. Please consult the contributing guide for more details.

> They barely have support for anything less. Note that 3 months ago it was at 1.6.0, as included in @dotlambda's [previous incarnation of this PR](https://github.com/NixOS/nixpkgs/pull/233393/files#diff-1e6bf80c8cd719f30a6f001e8aa296035de88b1cdb8840b953cd66b5f203a44eR35). Multiple breaking bugs and performance regressions were addressed since 1.5.1, making the version in-Nixpkgs currently dangerous to use.

What hexa means is that this PR should target staging. See my standard template on how to do that.

----

This PR rebuilds a lot of packages which means we must target staging. Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging) to not potentially ping a lot of people.

considerate commented on pkgs/tools/package-management/poetry/unwrapped.nix:
```diff
@@ -42,9 +43,36 @@
 }:
 
+let
+  poetry-core-overlay = poetry-core.overridePythonAttrs (old: rec {
+    version = "1.6.1";
+    src = fetchFromGitHub {
+      owner = "python-poetry";
+      repo = "poetry-core";
+      rev = version;
+      hash = "sha256-Gc22Y2T4uO39jiOqEUFeOfnVCbknuDjmzFPZgk2eY74=";
+    };
+    nativeCheckInputs = old.nativeCheckInputs ++ [
+      tomli
+      tomli-w
+    ];
+  });
+  poetry-plugin-export-overlay = poetry-plugin-export.overridePythonAttrs(old: {
+    version = "1.4.0";
+    src = fetchFromGitHub {
+      owner = "python-poetry";
+      repo = old.pname;
+      rev = "refs/tags/${old.version}";
+      hash = "sha256-okI91Z9u5w7IHpPb9jL4Hb8+MkYJEF2qm0mqqCdyKbk=";
+    };
+    nativeBuildInputs = ((builtins.filter (i: i.pname != poetry-core-overlay.pname) old.nativeBuildInputs) ++ [
+      poetry-core-overlay
+    ]);
+  });
```
I believe these overrides would be better to add to the python3 packageOverrides in the `default.nix` file:

https://github.com/NixOS/nixpkgs/pull/241458/files#diff-1e6bf80c8cd719f30a6f001e8aa296035de88b1cdb8840b953cd66b5f203a44eR8

I based this [commit](https://github.com/considerate/nixpkgs/commit/809696c8a33827dcba78d305c7449f85dc8242da) on this PR and I think it simplifies the build a bit compared to this PR.

Dessix replied:
I've incorporated your suggestions into the PR. I was unaware that the override would behave cyclically even back into the `poetry-plugin-export`'s `nativeBuildInputs`. This is far cleaner than the prior version- thank you ^^.

Dessix committed:
poetryPlugins.poetry-plugin-up: remove

Dropped support for python-plugin-up which has not been updated past poetry 1.4.2 after 1.5.x's significant change from tomlkit to tomli.
Poetry now includes a built-in update command.

considerate commented on pkgs/tools/package-management/poetry/unwrapped.nix:
```diff
@@ -84,12 +85,12 @@ buildPythonPackage rec {
     shellingham
     tomlkit
     trove-classifiers
+    tomli
+    tomli-w
```
I don't think `tomli-w` is used in `poetry`, only in `poetry-core`. Also, `tomli` should only be required if the python version is older than 3.11 because in `poetry` uses `tomllib` in 3.11.

https://github.com/python-poetry/poetry/blob/3fbd794ab8ab09bf99f90116c463996031b5544d/pyproject.toml#L59

considerate commented on pkgs/tools/package-management/poetry/default.nix:
```diff
@@ -1,39 +1,75 @@
-          hash = "sha256-GpZ0vMByHTu5kl7KrrFFK2aZMmkNO7xOEc8NI2H9k34=";
+          hash = "sha256-Gc22Y2T4uO39jiOqEUFeOfnVCbknuDjmzFPZgk2eY74=";
+        };
+        nativeCheckInputs = old.nativeCheckInputs ++ [
+          self.tomli
+          self.tomli-w
+        ];
```
```suggestion
        nativeCheckInputs = old.nativeCheckInputs ++ [
          self.tomli-w
        ];
```

It doesn't look like `poetry-core` depends on `tomli`, it only depends on `tomli-w`:
https://github.com/python-poetry/poetry-core/blob/0e16061b49decb4adf49df09309bc6d8f2c94e75/pyproject.toml#L42

considerate commented on pkgs/tools/package-management/poetry/unwrapped.nix:
```diff
@@ -84,12 +85,12 @@ buildPythonPackage rec {
   ] ++ lib.optionals (stdenv.isDarwin) [
     xattr
-  ] ++ lib.optionals (pythonOlder "3.11") [
-    tomli
```
Please revert to using this optional dependency for `tomli`.

considerate reviewed (approved):
Looks good to me!

SuperSandro2000 commented:
can you check why ofborg fails?

Dessix commented:
@SuperSandro2000
> can you check why ofborg fails?

Looks like https://github.com/NixOS/nixpkgs/pull/241227 may be the culprit. Hard to pick apart from how ofborg logs are structured for parallel tasks.

mirkolenz commented:
Poetry `1.4.2` currently also fails to build on the unstable channel due to the linked issue #241227 with the following message:

```text
> ERROR: Could not find a version that satisfies the requirement cachecontrol[filecache]<0.13.0,>=0.12.9 (from poetry) (from versions: none)
> ERROR: No matching distribution found for cachecontrol[filecache]<0.13.0,>=0.12.9
```

K900 commented:
```
diff --git a/pkgs/tools/package-management/poetry/default.nix b/pkgs/tools/package-management/poetry/default.nix
index 48263c1f309..913b95fc8ef 100644
--- a/pkgs/tools/package-management/poetry/default.nix
+++ b/pkgs/tools/package-management/poetry/default.nix
@@ -55,12 +55,12 @@ let
           self.time-machine
         ];
       });
-      poetry-plugin-export = super.poetry-plugin-export.overridePythonAttrs(old: {
+      poetry-plugin-export = super.poetry-plugin-export.overridePythonAttrs(old: rec {
         version = "1.4.0";
         src = fetchFromGitHub {
           owner = "python-poetry";
           repo = old.pname;
-          rev = "refs/tags/${old.version}";
+          rev = "refs/tags/${version}";
           hash = "sha256-okI91Z9u5w7IHpPb9jL4Hb8+MkYJEF2qm0mqqCdyKbk=";
         };
       });
```

Is probably what was intended here?

dotlambda commented on pkgs/tools/package-management/poetry/default.nix:
```diff
@@ -1,39 +1,74 @@
+          self.time-machine
+        ];
+      });
+      poetry-plugin-export = super.poetry-plugin-export.overridePythonAttrs(old: {
```
`poetry-plugin-export` should be removed from `python3Packages`.

Hamled commented:
> The following patch should fix it:
> 
> ```
> +  patches = [
> +    (fetchpatch {
> +      url = "https://github.com/python-poetry/poetry/pull/8055/commits/a5a6df8d8792a8ae47f1d1b57b439dee3ee4d64f.patch";
> +      hash = "sha256-c2IYrdc2RaGZupRH2n26Rr2eUpDPV87Aw/P3x54j6ts=";
> +    })
> +  ];
> ```

I wasn't able to get this to apply to version 1.4.2 as it was originally made for 1.5.0 I believe. I made a patch has the same changes backported to 1.4.2, but it is untested beyond building the package and running `poetry init`:
```
+  patches = [
+    (fetchpatch {
+      url = "https://github.com/Hamled/poetry/commit/d2802deec588a07626e89da33ad41bb36b885426.patch";
+      hash = "sha256-8Yznjvwz+kg4U+kml/GbodHAKYnvEn86TZfD65j/86A=";
+    })
+  ];
```

dotlambda commented:
I updated https://github.com/NixOS/nixpkgs/pull/233393 to build poetry-plugin-up with support for poetry 1.5.

fabaff commented on pkgs/tools/package-management/poetry/default.nix:
```diff
@@ -1,39 +1,74 @@
     packageOverrides = self: super: {
       poetry = self.callPackage ./unwrapped.nix { };
 
+      filelock = super.filelock.overridePythonAttrs (old: rec {
```
Would not be needed with https://github.com/NixOS/nixpkgs/pull/242402
